Add support for pushing to GitLab using merge requests.

maxking opened the pull request:
This support is based on "lab" CLI, which is inspired by "hub" tool, which is
used for GitHub's pull request support.

"lab" also proxies a lot of commands to "hub", so parts of the Github and
Gitlab's requirements are abstracted out in a mixin class and then only the
basic functionality that is different is implemented via separate classes.

Creating a draft PR for early reviews, will update with test cases soon.

pep8speaks commented:
Hello @maxking! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`weblate/vcs/git.py`](https://github.com/WeblateOrg/weblate/blob/b1771c0584203998328f1ac73b9d26c4ae1b87cf/weblate/vcs/git.py):

> [Line 590:5](https://github.com/WeblateOrg/weblate/blob/b1771c0584203998328f1ac73b9d26c4ae1b87cf/weblate/vcs/git.py#L590): [E301](https://duckduckgo.com/?q=pep8%20E301) expected 1 blank line, found 0

nijel commented on weblate/vcs/git.py:
```diff
@@ -615,7 +590,39 @@ def push(self):
+
+    name = 'GitHub'
+    _cmd = 'hub'
+    _username = settings.GITHUB_USERNAME
```
This assigns a value to local variable at class initialization, what makes it impossible to override setting in the tests. It might be better to assign setting name here and use getattr to fetch actual value later.

Apparently running tests for GitHub pull requests was broken for some time, I've just fixed it in bbaadb62c696442c380b7c888173047ee7bc7e2b.

maxking replied:
Ah, that makes sense.

How about I convert `_username` to a `@property` instead?  It is slightly smaller and cleaner than multiple `getattr(settings, self._username_setting_name)`.

nijel replied:
Good idea.

nijel commented:
Looks like you've done something wrong in rebasing, the merge request contains lot of upstream commits...

maxking commented:
Yeah, I realized that. Will need to cherry-pick the commits out of this mess.

nijel commented on weblate/vcs/git.py:
```diff
@@ -544,57 +544,32 @@ def push(self):
         """Push given local branch to branch in forked repository."""
-        if settings.GITHUB_USERNAME:
-            cmd_push = ['push', '--force', settings.GITHUB_USERNAME]
+        if self._username:
```
I know this code was originally there, but can you please remove this condition? It's not useful anymore (I'm too lazy to dig out original use case in git history). The class will not be used without username being available.

maxking replied:
Sure, will do!

nijel replied:
I've done this in 3d7cc2ddf2, you will have to rebase, sorry.

maxking replied:
Done.

nijel commented on weblate/vcs/git.py:
```diff
@@ -544,57 +544,32 @@ def push(self):
 
 
-class GithubRepository(GitRepository):
+class _GitRemoteRepositoryMixin(GitRepository):
```
This is not really a mixin, but rather a base class, so I'd prefer cleaner name. How about following?
```suggestion
class GitMergeRequestBase(GitRepository):
```

maxking replied:
Sounds good, do you need a `_` in the beginning since this is kind of an Internal Implementation detail? :)

nijel replied:
It's still base class and somebody else might want to build third party vcs backend on that.

nijel commented on docs/vcs.rst:
```diff
@@ -340,3 +340,66 @@ monolingual translations).
+GitLab
+------
+
+.. versionadded:: 3.10
```
I think we can do it for 3.9 :-)

maxking replied:
Ah, I didn't check the latest version. Will do.

nijel replied:
```suggestion
.. versionadded:: 3.9
```

maxking commented:
I haven't tested it out yet on a real environment, part of which is because last time I tried to start up a local instance using the `rundev.sh` script, it kept on failing due to missing container dependencies. I haven't had time to actually get to run Weblate locally yet. 

I probably need to read the docs more for the setup from sources and it should be good to go, but I hoped to skip the setup with containers if possible.

As soon as I can get this to run locally and verify against a repo, I'll remove the draft status. Thanks for patiently reviewing everything :)

nijel commented:
Hmm, the `rundev.sh` was contributed some time ago and I don't think anybody uses that now, so it can be broken. I fixed the most visible breakage in 365ee4680dd1e8e9e197532a6092c752a287afda

maxking commented:
I think I was finally able to test the MR creation (https://gitlab.com/maxking/hyperkitty/merge_requests/19) and seems to work. There were a few problems with the original code, which I fixed.

Some interesting details I discovered:
- `lab` does not allow specifying which branch to use as the source and defaults to the current branch, because of which, I had to wrap around the MR creation with a checkout of the remote branch we want to use as the source.
- `lab` does not have a way to specify the configs via the environment variables, so it must be created in the `DATA_DIR/home/.config/lab.hcl`. I have updated the docs to setup lab as well, so it should reflect the same changes.

I think I have fixed all of your suggestions as well, but please let me know if you have any more.

I am really excited to be able to use this feature in hosted.weblate.org and retire my custom shell script running on cron :)

nijel commented:
Thanks, can you please check the test failure? There seems to be something passing None to the commandline:

```py-tb
Traceback (most recent call last):
  File "/home/travis/build/WeblateOrg/weblate/weblate/vcs/tests/test_vcs.py", line 234, in test_upstream_changes
    self.add_remote_commit()
  File "/home/travis/build/WeblateOrg/weblate/weblate/vcs/tests/test_vcs.py", line 151, in add_remote_commit
    repo.push()
  File "/home/travis/build/WeblateOrg/weblate/weblate/vcs/git.py", line 558, in push
    self.push_to_fork(self.branch, self.branch)
  File "/home/travis/build/WeblateOrg/weblate/weblate/vcs/git.py", line 539, in push_to_fork
    self.execute(cmd_push + ['{0}:{1}'.format(local_branch, fork_branch)])
  File "/home/travis/build/WeblateOrg/weblate/weblate/vcs/base.py", line 191, in execute
    args = [arg.encode('utf-8') for arg in args]
  File "/home/travis/build/WeblateOrg/weblate/weblate/vcs/base.py", line 191, in <listcomp>
    args = [arg.encode('utf-8') for arg in args]
AttributeError: 'NoneType' object has no attribute 'encode'
```

maxking commented:
@nijel I think the tests are passing now, there are some jobs which might need re-starting since they seems to have errored out due to something getting stuck.

nijel commented:
We're hitting the 50 minutes job limit on Travis CI, right now we're testing Drone CI as a replacement for that and the tests passed there.

nijel commented:
Merged, thanks for your contribution!
